### PR TITLE
Implement release plugin

### PR DESCRIPTION
**Pull request by daymxn**
Per [b/317081045](https://b.corp.google.com/issues/317081045),

This implements a new plugin for preparing a release. This plugin provides functionality to;

- Update the project version declared in the `gradle.properties` file
- Create a new `.api` file in the *new* `api` directory
- Automatically generate the release notes

This PR also fixes the following:

- [b/319482236](https://b.corp.google.com/issues/319482236) -> Store released `.api` files in a `api` directory

> [!WARNING]
>
> This PR should not be merged until the following is complete:
>
> - [x] The branch is not currently in a release state
> - [x] All of the previous release's `.api` files are present in the `api` directory

### CODE REVIEW

File: plugins/src/main/java/com/google/gradle/tasks/VersionBumpTask.kt [hunk 1]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2023 Google LLC
```
Comment (ryanwilson) [inline plugins/src/main/java/com/google/gradle/tasks/VersionBumpTask.kt:2]: Nit: 2024

File: .github/workflows/check_for_api_changes.yml [hunk 2]
```diff
@@ -36,7 +36,7 @@ jobs:
           cache: gradle
 
       - name: Copy saved api to branch
-        run: mv ~/released.api generativeai/released.api
```
Comment (rlazo) [inline .github/workflows/check_for_api_changes.yml:39]: public?

File: .github/workflows/check_for_api_changes.yml [hunk 3]
```diff
@@ -18,12 +18,12 @@ jobs:
           distribution: temurin
           cache: gradle
 
-      - name: Update released.api
+      - name: Export the api file
```
Comment (rlazo) [inline .github/workflows/check_for_api_changes.yml:21]: Generate

File: plugins/src/main/java/com/google/gradle/plugins/ReleasePlugin.kt [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+import com.google.gradle.util.readFirstLine
+import java.io.File
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+import org.gradle.kotlin.dsl.named
+import org.gradle.kotlin.dsl.provideDelegate
+import org.gradle.kotlin.dsl.register
+
+/**
+ * A Gradle plugin for preparing the release of a [Project].
```
Comment (rlazo) [inline plugins/src/main/java/com/google/gradle/plugins/ReleasePlugin.kt:34]:
```suggestion
 * A Gradle plugin for preparing a release.
```

File: plugins/src/main/java/com/google/gradle/tasks/CopyFileTask.kt [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+import org.gradle.api.provider.Property
+import org.gradle.api.tasks.Copy
+import org.gradle.api.tasks.InputFile
+import org.gradle.api.tasks.OutputFile
+import org.gradle.api.tasks.TaskAction
+
+/**
+ * Copies a file (or directory) from one place to another.
+ *
+ * An alternative to the standard [Copy] task provided by gradle; that allows better interop with
```
Comment (rlazo) [inline plugins/src/main/java/com/google/gradle/tasks/CopyFileTask.kt:30]: How is it better at those than the standard task?

File: .github/workflows/check_for_api_changes.yml (hunk 2 above)
Comment (daymxn) [reply .github/workflows/check_for_api_changes.yml:39]: sgtm

File: plugins/src/main/java/com/google/gradle/tasks/CopyFileTask.kt (hunk 5 above)
Comment (daymxn) [reply plugins/src/main/java/com/google/gradle/tasks/CopyFileTask.kt:30]: Allows you to more easily specify a specific file to export to. The biggest issue though is that the standard copy task doesn't let you call `rename` with a `Provider`- which makes copying an individual file to a specific location/file a bit difficult when working with task providers and the like. The `Copy` task honestly needs an update from Gradle.

**Review by rlazo — APPROVED** [review]

**Pull request merged**